Bringing ndjson(document_stream) to On Demand

NicolasJiaxin opened the pull request:
I have trying to implement just a simple document stream for On Demand. The main issue, as discussed by @lemire in this [comment](https://github.com/simdjson/simdjson/issues/1464#issuecomment-870690949), is that On Demand does not know where the end of a single document lies in a document stream (if stage1 covers multiple documents). To overcome this, I created a JSON iterator that will be used to traverse a document and go to the next document when needed. This JSON iterator is also used to create a document instance when the *operator is called.

- [x] implements threaded version so that stage 1 is processed independently
- [x] add the rest of the tests so that we have as good a coverage as the DOM document_stream, 
- [x] add documentation and examples,
- [x] add benchmarking.

Fixes https://github.com/simdjson/simdjson/issues/1464

NicolasJiaxin commented on include/simdjson/generic/ondemand/document_stream-inl.h:
```diff
@@ -0,0 +1,179 @@
+inline void document_stream::next() noexcept {
+  // We always enter at once once in an error condition.
+  if (error) { return; }
+  /** This does not work, but this is the idea
+  do {
+    iter.skip_child(0);
+  } while(iter._depth != 0);
+  iter._depth = 1;
+  if (error) { return; }
+  */
```
This is the code that I would like to implement, but for some reason, it does not let me call skip_child on iter (Segmentation fault)

NicolasJiaxin replied:
It seems that this is caused by this line in `skip_child` 
```c++
auto end = &parser->implementation->structural_indexes[parser->implementation->n_structural_indexes];
```

NicolasJiaxin commented:
This should let us at least get the next document in the stream for now (see `ondemand_document_stream_tests`). However, the document instance from the *operator only works properly when called on the first document.

lemire commented:
@NicolasJiaxin Excellent first start.

lemire commented:
@NicolasJiaxin Can you try to compile your PR with `SIMDJSON_EXCEPTIONS=OFF`. This is annoying but we try to support the exception-free mode and I think your code does implicit conversions from simdjson_result<T> to T some of the time. (This happens to me too all of the time.)

NicolasJiaxin commented:
Yes, indeed. It happens with the *operator, converting from simdjson_result<document> to document.

NicolasJiaxin commented:
I am not sure how to fix it though..

jkeiser commented:
My plane is about to lift off, but you can get a much better idea of where you are first running into trouble with a debug build (you have to rm -rf your build directory first, then `cmake -DCMAKE_BUILD_TYPE=Debug ..` (plus any other arguments you want to pass).

The level above the document should be depth=0, btw. When we step into a document (or start the stream), the depth is 1.

jkeiser commented:
It is not impossible I added an assert in skip child that the target depth be greater than 0; if that's the case you are going to get weird results in a release build, because SIMDJSON_ASSUME tells the compiler to assume things are true without checking them in release. Debug builds will check and assert.

NicolasJiaxin commented:
thanks @jkeiser! I think the faulty line is the one I mentioned earlier. This one:
```c++
auto end = &parser->implementation->structural_indexes[parser->implementation->n_structural_indexes];
```
For some reason, trying to access the structural indexes of the parser field causes a segmentation fault.

NicolasJiaxin commented:
For now though, I have gotten around this by doing the same algorithm manually, which lets us move the json_iterator to the next document.

lemire commented:
@NicolasJiaxin I am going to take your code out for a spin. I'll comment back soon.

NicolasJiaxin commented:
The problem seems to be with the *operator. If not called, everything seems fine. We can even call skip_child properly. But after it is called, things break. Structural indexes and addresses are messed up.

lemire commented:
I am working on it. I will have something soon. 

Feel free to go out and celebrate Canada day!

lemire commented:
I have now made it where you stopped.

lemire commented:
@NicolasJiaxin Conceptually, there is a problem here:

```C++
simdjson_really_inline simdjson_result<ondemand::document> document_stream::iterator::operator*() noexcept {
  // Note that in case of error, we do not yet mark
  // the iterator as "finished": this detection is done
  // in the operator++ function since it is possible
  // to call operator++ repeatedly while omitting
  // calls to operator*.
  if (stream->error) { return stream->error; }

  return document::start(std::move(stream->iter));
}
```

After this call, `stream->iter` is meant to be discarded. So I don't think we want to do it that way.

lemire commented:
Let me correct myself: this code is fine if we later recover ownership.

lemire commented:
This is tricky.

1. In our model, the document owns the json_iterator and all that other folks can do is have a pointer to it.
2. We would like the document_stream to produce document instances but we can't have two owners of the json_pointer.

lemire commented:
@NicolasJiaxin @jkeiser 

I think I have one possible way out, and I took the liberty to push it directly into @NicolasJiaxin's fork so that this PR is updated.

- Nicolas had the document_stream own the json_iterator, and then he would pass the json_iterator to a document's constructor. However, the document's constructor would move the json_iterator in the document, and so that document_stream would be left with a broken/invalid json_iterator.
- So I tried something else. The document_stream owns a document, that we reuse. The document itself owns the json_iterator. The counterpart is that we must return a *reference* to the document to the user, the user can't have a document instance, we get to own it.
- To make it work, I had to add an attribute to json_iterator, which is the 'root' or starting point. Maybe it is not the right place, but someone must keep track of where the current document begins.

I made a nasty (broken) hack: I disabled `peek_last()` checks the document's code. This breaks stuff but allows me to run `ondemand_document_stream_tests`.

The implementation is full of holes and needs much more love...

But I am hoping we can make a bit more progress.

What is crucially important here is that the current code must have a clear line of ownership for the json_iterator as there is a single instance (this makes the code more challenging to write, but it was deliberate).

Let us chat!!!

NicolasJiaxin commented:
It looks very good!! I also had the idea to do make the stream own a document, but I was not able to make it work. I do think though that it is a good idea. As for the *operator, I am not sure how "limiting/different" it is to return a reference to the document instead of providing document instance. From what I can see, we cannot use the `auto` keyword because the deleted copy constructor will be implicitly called, but I am not very knowledgeable in C++. I think the cleanest implementation would be to return a document instance. 

As for the starting point/root of the current doc, I think implementing it in the json_iterator class is a good idea. It does not break single document parsing, and it removes the assumption that the start is always at the start of the structural indexes (which would be problematic if we tried to keep track of the start/root from the document_stream class). This means more flexibility that could be used elsewhere than in the document_stream.

NicolasJiaxin commented:
Also, just as a line of thought, we could always try to continue with the json_iterator idea if we returned a document with a distinct json_iterator. This involves some implications. First, we would have to update its position correctly to point at the start of the current document. I think we would also need to make use of the attribute `root` that you added.

The pro that this approach has, is that it allows us to return a document instance directly. However, I think this approach has several cons (hence why I think yours is probably better). First, this could use a lot of memory since the document instances are completely separated from the document_stream. Secondly, both json_iterators (from the document stream and the document instance) would have the same parser. I don't know exactly how it could be troublesome, but it does look like it could create bugs and a lot of trouble. Finally, there could be some performance issue. Since the json_iterator from the document_stream is "untouched", everytime we call `next()`, we would have to start at the beginning of the document to skip it. 

This is probably not a very good approach compared to document one, but I just want to mention it to keep it in mind *just in case*.

lemire commented:
> From what I can see, we cannot use the auto keyword because the deleted copy constructor will be implicitly called, but I am not very knowledgeable in C++. I think the cleanest implementation would be to return a document instance.


I think `auto&` will work. See this example:

```C++

struct T {
    T() {}   
     T(const T & x) = delete;
};

T& f() {
    static T x;
    return x;
}

void g() {
    auto& z = f();
}
```

I would also prefer to return a document instance, but it would require changing a lot of the existing design. As long as we have document instances which own the json_iterator instance, then I do not see how we can efficiently return a document instance. Maybe I am just not seeing it, but every other option I considered was more complicated and error prone.


> As for the starting point/root of the current doc, I think implementing it in the json_iterator class is a good idea. It does not break single document parsing, and it removes the assumption that the start is always at the start of the structural indexes (which would be problematic if we tried to keep track of the start/root from the document_stream class). This means more flexibility that could be used elsewhere than in the document_stream.

Right.

At this point, I suggest you continue with the current direction... I am sure that @jkeiser and others will have better ideas, but we can still move forward and correct the design later if needed.

Try to get more functionality going. Don't necessarily try to solve all problems, just move it forward and next week we can see where we are at.

Hint: compile with `SIMDJSON_EXCEPTIONS=OFF`. Disabling exception support will reduce the number of CI errors.

lemire commented:
One thing that is daring about the new approach we do not return a `simdjson_result<document>` instance from the iterator. Returning `document&` is conceptually fine, I expect... this will work...

```C++
for (auto & doc: stream) {
...
}
```

But we are diverging from the DOM ndjson implementation in how we handle errors. Maybe it will turn out fine (and maybe even better/simpler), but I would have preferred to stick to the mature approach.

In some sense, there is an underlying drastic change because in the DOM approach, one you got the document, it was fully validated. So you either got an error when you got the document or none at all.

With the On Demand approach, conceptually, getting a 'document' means little in terms of error handling. The document could be all wrong. You just have a pointer a possible document.

So I think that this direction shows promise... but we will have to be careful and write a fair amount of tests.

NicolasJiaxin commented:
Ok, I will continue in this direction. I will report here anything that is problematic.

NicolasJiaxin commented:
> With the On Demand approach, conceptually, getting a 'document' means little in terms of error handling. The document could be all wrong. You just have a pointer a possible document.
> 
> So I think that this direction shows promise... but we will have to be careful and write a fair amount of tests.

Yes, it seems like that. Even though there are technically less steps with On Demand, we have to be wary that everything works as intended.

NicolasJiaxin commented:
Is it okay if I move `DEFAULT_BATCH_SIZE`, `MINIMAL_BATCH_SIZE` and `MINIMAL_DOCUMENT_CAPACITY` from `dom/parser.h` to `common_defs.h`?

lemire commented:
@NicolasJiaxin I would suggest duplicating them. Given that we have namespaces, there should be no conflict. And the additional bloat is tiny.

The benefit in duplicating them is that we can adjust them with only a local effect. For example, DEFAULT_BATCH_SIZE might need tweaking that is specific to our new implementation.

NicolasJiaxin commented:
I'm currently trying to add the feature to load new batches when required, but there seems to be a problem with `run_stage1`. For some reasons, it advances the json_iterator forward by 2 tokens when loading a new batch (initial batch is not affected by this issue). I have included print statements to clearly illustrate this when running `ondemand_document_stream`. Is this caused by the lack of padding when the second batch is loaded?
Also, note that I changed `MINIMAL_BATCH_SIZE` to 4 for testing purposes.

lemire commented:
I'll have a look Monday!

NicolasJiaxin commented:
Yes of course. Enjoy your week-end!

jkeiser commented:
I'm committing the cardinal sin of not having read everything here before commenting, but my preliminary thoughts around document_stream were to make document_stream the sole owner in all cases:

* `ondemand::document_stream` would yield a `document_stream::document` instead of an owned `ondemand::document`.
* `document_stream::document` would be just like `value` (possibly even containing one) in that it doesn't own anything.
* `ondemand::document` would own a `document_stream`.
* `ondemand::documnt_stream` would be the only class to directly own a `json_iterator`.

Just partially-fleshed-out suggestions without the context of fully grokking what you guys wrote yet!

lemire commented:
@jkeiser I think that what you propose would work and might be the right design in the end. However, I am concerned that it would make this PR invasive so I would suggest that we continue with the current path, and the come back for a second pass.

If you believe that your proposal could be implemented with little code, please express your disagreement with my stance.

The "return a document reference" approach is not a pretty design, but my current feeling is that it has the benefit of introducing almost no change to our code base and to be implementable with really little code. (I may yet be proven wrong. It is not implemented yet... I am only guessing that we are "kind of close to a working system".)

What you suggest may require introducing a sort of pseudo-document class, and it would require changing the document class so that it refers to a document_stream class... That requires touching the core of the On Demand front-end... Not something I am eager to do while we are adding new functionality. Intuitively, a redesign better done on its own (as part of PR that does *just that*).

lemire commented:
@NicolasJiaxin I am looking at the issue now. Updates will follow.

lemire commented:
> I'm currently trying to add the feature to load new batches when required, but there seems to be a problem with run_stage1. For some reasons, it advances the json_iterator forward by 2 tokens when loading a new batch (initial batch is not affected by this issue). I have included print statements to clearly illustrate this when running ondemand_document_stream. Is this caused by the lack of padding when the second batch is loaded?

Note that I do not see this.

Still working.

NicolasJiaxin commented on include/simdjson/generic/ondemand/document_stream-inl.h:
```diff
@@ -0,0 +1,183 @@
+      batch_start = next_batch_start();
+      if (batch_start >= len) { break; }
+
+      std::cout << "FINE: " << doc.iter.peek() <<std::endl;
+      error = run_stage1(*parser, batch_start);
+      std::cout << "BROKEN: " << doc.iter.peek() <<std::endl;
```
You should get print messages from these lines when running doc_index in ondemand_document_stream_tests @lemire

lemire commented:
I do however see the bug. Working on it.

lemire commented:
@NicolasJiaxin 

> You should get print messages from these lines when running doc_index in ondemand_document_stream_tests @lemire


Yes. But they print what you'd expect (e.g., `[1,2,3]`). It does not matter though because I see the issue.

It is just that there are many, many moving parts and the On Demand front-ends makes a lot of assumptions about how things are. 

I am working on coming up with a clean way out.

The core issue is that you can't just run stage 1 and expect things to work out. This worked in DOM because we updated the pointers. With the On Demand front-end, we have pointer references all over that need to be updated.

lemire commented:
@NicolasJiaxin Ok. I have a fix coming up. Give me 5 to 10 minutes.

lemire commented:
@NicolasJiaxin There we go. I think this commit will help...

https://github.com/simdjson/simdjson/pull/1643/commits/c44167f598f58a90124fce06a0385507141090b1

Please review it.

lemire commented:
@NicolasJiaxin As the failed tests suggest, we still have work to do. But one step at a time.

NicolasJiaxin commented on tests/ondemand/ondemand_document_stream_tests.cpp:
```diff
@@ -0,0 +1,62 @@
+
+namespace document_stream_tests {
+
+    std::string my_string(ondemand::document& doc) {
+        std::stringstream ss;
+        ss << doc;
+        return ss.str();
+    }
```
Is it possible that my_string actually consumes the document?

lemire replied:
I expect it does, yes.

lemire replied:
So when I changed your tests by adding to_string everywhere, I may have made it easier to pass in some sense. But it should be easy enough to add more tests.

NicolasJiaxin commented:
@lemire The resync looks good! I will continue to work on functionality, and then work on making this PR pass the tests. Thank you for the help

NicolasJiaxin commented:
I've added a `source` function similar to the one in the DOM implementation. For now, it passed the test I have written, but there is more testing to do if we want to keep this function.

NicolasJiaxin commented:
Also, just a question: why is the json iterator initialized with depth=1 even though we have not consumed the initial structural index?

lemire commented:
> Also, just a question: why is the json iterator initialized with depth=1 even though we have not consumed the initial structural index?

@jkeiser : as the depth expert... you want to take this question?

lemire commented:
@NicolasJiaxin 

> I've added a source function similar to the one in the DOM implementation. For now, it passed the test I have written, but there is more testing to do if we want to keep this function.

We may be able to replace it with https://github.com/simdjson/simdjson/pull/1642 in the future (so that we have less code), but it sounds good at a glance.

lemire commented:
@NicolasJiaxin I pushed a minor commit to your fork. Please sync.

lemire commented:
@NicolasJiaxin I do not plan to keep on pushing stuff on your fork, but I also did not want you left with silly puzzles. :-)

NicolasJiaxin commented:
Yes thank you. I am starting to work on making this PR pass all the tests. Then, I will add more tests so we have better coverage. And after that, I think we can try to add the multithread feature.

jkeiser commented:
Depth is always the depth of the current, unconsumed token. We start at 1 because the first token is a value; when we finish with that value, depth will decrease to zero.

lemire commented:
@NicolasJiaxin Try syncing with master as I just solved an "infinite loop" bug that I had introduced earlier.

NicolasJiaxin commented:
I've added `error()` to the iterator class. This might not be the best place to put it though (might be better in the `document_stream` class). I just wanted to add a way to retrieve any error in the streaming. In the DOM approach, this was done via the *operator which returned the document or an error. However, for this implementation, we are returning a reference to a document and not a simdjson_result. Is this something we can/want change so that we can get the error the "usual" way?

lemire commented:
>  This might not be the best place to put it though

It is fine to make such 'temporary' choices. We can then revisit the design once everything is solid.

NicolasJiaxin commented:
I fixed `source()` because it used to return the whitespaces between documents in the string_view. Now, it should be fixed for documents that are either objects or arrays. However, it still includes whitespaces for single scalar value documents, but since this feature might get replaced by #1642 in the future, I don't think this detail is very significant. If `source()` becomes more relevant, I will fix this issue. For now, it is mainly for testing and debugging.

lemire commented:
@NicolasJiaxin I am hoping we can use PR https://github.com/simdjson/simdjson/pull/1651 "eventually" as it would give us plenty of flexibility while simplifying the API but, even so, it could be that your approach has benefits *and* it helps debugging. So good move...

White space before or after the document is quite fine.

lemire commented:
@NicolasJiaxin There was a CI issue which I hopefully fixed with a small commit (which should not interfere with your work).

NicolasJiaxin commented:
Thank you! By the way, I am working on the resync that you did, because the iter._root is not updated correctly. It is put back at the beginning, and so the root points at the first structural index of the first document. Does calling stage_1 reset the structural indexes?

NicolasJiaxin commented:
The problem is that parser->implementation->structural_indexes.get() does not change. It is still at pointing at the first structural index, even after calling stage 1.

lemire commented:
>  Does calling stage_1 reset the structural indexes?

Ok. So if you have a parser instance, it owns an array which contains a "structural index": it is a list of 32-bit integers which contain offsets and provide easy access to the structural and pseudostructural elements. So if you ask it to index a memory buffer containing the string `{ 1, }` then it will have the values 0, 2, 3, 5 corresponding respectively to `{`, `1`, `,` and `}`.

Each time you call stage 1 on the parser, it recomputes the indexes based on the provided buffer.

Now, if you divide your document into windows, then these 'structural indexes' will always start at offset 0 *in the window*. The stage1 routine has no concept of window. As far as it is concerned, you provide a view of the whole thing.

But the json_iterator reads data from a pointer (I think it is called buf) as well as these offsets (structural indexes). The two must be in sync. After you call stage 1, you get new indexes, but you must also start accessing the data from *within the indexed window*.

To answer your question... yes, stage 1 recomputes entirely the indexes and the whole ones are gone (within this parser instance).

>  By the way, I am working on the resync that you did, because the iter._root is not updated correctly. It is put back at the beginning, and so the root points at the first structural index of the first document.

After calling stage 1 at some point, I did the following...

```
doc.iter = json_iterator(&buf[batch_start], parser);
```

That should do what I described above needs doing. It moves the pointer to `&buf[batch_start]` which should point at the start of the new window, and it takes in the parser instance.

Something like this needs to be done *each time* you call stage 1.

It should initialize `_root` in the following manner...

```
_root{parser->implementation->structural_indexes.get()},
```

So `_root` should point at the first structural index from the *new* indexes. So if `parser->implementation->structural_indexes.get()` points at `3`, say, then the first structural element should be at `&buf[batch_start+3]`.

Of course, maybe I messed up.

NicolasJiaxin commented:
@lemire That makes sense, thanks for the clear explanation. I think I might just have forgotten the offset in `buf`. I will try to offset it with `batch_start`.

lemire commented:
> The problem is that parser->implementation->structural_indexes.get() does not change. It is still at pointing at the first structural index, even after calling stage 1.

Ok. So `parser->implementation->structural_indexes.get()` only contains 32-bit offsets. If you print its values before and after calling stage 1, it should definitively be different (it is possible that the first offset remains zero but that would only indicate that you have a structural element at the beginning of each window).

Saying that "It is still at pointing at the first structural index" is probably incorrect. You can only interpret the offset in question correctly when you also have the matching pointer, and the pointer in question needs to be updated so that it points at the beginning of the new window.


BTW this is not easy code. It is highly abstract code that has many moving parts at once.

And wait until you parallelize it... then it gets really mind boggling.

But once you make it work, it will be stupidly powerful. :-)

NicolasJiaxin commented:
> Saying that "It is still at pointing at the first structural index" is probably incorrect. 

It was indeed incorrect. I was able to fix it now. I thought something might have been wrong with the resync, but it was actually my fault, `source()` did not account for `batch_start`.

> But once you make it work, it will be stupidly powerful. :-)

Can't wait to see those benchmarks :-)

NicolasJiaxin commented:
I implemented the multithread version in the same way as it was for DOM: having a parser in the background that loads one batch in advance, and then swap to that parser when needed so that we always have a fresh batch prepared. However, I was not able to create a field `ondemand::parser stage1_thread_parser` (background parser) and had to use a parser pointer instead (even when I included the header file `parser.h` before `document_stream.h` to avoid incomplete type issue). I also did a bad hack by manually creating a parser instance and using it to initialize the pointer during the tests. I did so to see if the tests would succeed if we had a second parser in the background.

NicolasJiaxin commented:
bd546e1 is the hacky version that is "working", while 25de329 would be the "correct" way of implementing, but that has an error when compiling.

lemire commented:
@NicolasJiaxin I plan to review early tomorrow. 😅

NicolasJiaxin commented:
@lemire Yes no problem!

lemire commented:
Note that there is a scheduled power outage at my home tomorrow and I still can't return to my campus office, so I might be less efficient than I would otherwise be.

lemire commented:
@NicolasJiaxin You probably understand the issue at this point, but just to be sure... The compiler is telling you that you are trying to create a copy of a parser instance. By design, we do not let you do that. By copy, I mean calling a copy constructor (not a move constructor).

I'll propose a fix soon. (Need to read more of the code.)

NicolasJiaxin commented:
Yes, I understood that from the error message, but I do not understand why is the copy constructor called. It seems that even just by adding a parser instance field without any initialization, I get in trouble.

lemire commented:
@NicolasJiaxin Yes. I will soon propose a fix.

lemire commented:
@NicolasJiaxin Have a look at commit https://github.com/simdjson/simdjson/pull/1643/commits/305ef7d56c92c34da4526d72ba9c4d68067feca1

I think that the issue had to do with the fact that we did not provide a move assignment for the ondemand::parser class. I think clang would do it automatically for us, but other compilers would not.

The net result was that you did not have a move assignment, so the compiler would try a default assignment, but it was deleted, so it would just fail violently.

My single-line change should help. It will also help, in practice, people who need to work with ondemand::parser instances.

This is, btw, one motivation for the work that you are doing right now. You are exposing issues with the On Demand front-end by stress testing it. So even people who will never touch document_stream should benefit from your work, in the end.

Note that, at least on my machine, I get test errors... so it is not the end of the road.

lemire commented on include/simdjson/generic/ondemand/document_stream-inl.h:
```diff
@@ -0,0 +1,406 @@
+  worker->finish();
+  // Swap to the parser that was loaded up in the thread. Make sure the parser has
+  // enough memory to swap to, as well.
+  this->worker->stage1_thread_parser = & this->stage1_thread_parser;
```
I do not like this line.... `this->worker->stage1_thread_parser = & this->stage1_thread_parser;`.

The DOM version of this code looks better...

```
std::swap(*parser, stage1_thread_parser);
```

I am not sure why your version is more complicated.

Be mindful that complicated code is harder to check and more error prone.

NicolasJiaxin replied:
I was having with making swap work, and this is the way I found it to make it work, sorry I forgot telling about that.

lemire replied:
@NicolasJiaxin It should not be necessary. If it is, let us investigate it.

lemire commented:
Oh fun!!! The thread sanitizer reports a data race!!!

lemire commented:
@NicolasJiaxin Make sure you try to build with `SIMDJSON_SANITIZE_THREADS=ON`. (I recommend using just one sanitizer at any one time, and consider building in debug mode.)

Ping me if you get stuck.

NicolasJiaxin commented:
OK, thank you a lot! I will continue trying

lemire commented:
@NicolasJiaxin  At least now the code is building.

NicolasJiaxin commented:
The biggest issue was actually with the `std::swap` that you mentioned. I changed it back to the DOM version. I think the issue I had with it earlier was also related that to the fact there was no move assignment for parser instances. Could that be the case?

NicolasJiaxin commented:
Also, I am unsure why the old line `this->worker->stage1_thread_parser = & this->stage1_thread_parser;` caused a data race.. It is not like it actually notify the thread to start working yet. Could it be that there is an issue with the mutex? Is it something I should look into?

lemire commented:
@NicolasJiaxin I am highly suspicious of the data-race sanitizer. It appears to have false positives. You can find ample comments that indicate my concerns. Nevertheless, we seek to get a clear bill of health from the sanitizers when we can.

NicolasJiaxin commented:
@lemire are we ready to start benchmarking? If everything else is fine, of course.

lemire commented:
@NicolasJiaxin Yes. 

Can you also update `basics.md`... making your work the new default. 

https://github.com/simdjson/simdjson/blob/master/doc/basics.md#newline-delimited-json-ndjson-and-json-lines

lemire commented on doc/basics.md:
```diff
@@ -933,60 +933,38 @@ format. If your JSON documents all contain arrays or objects, we even support di
-dom::document_stream docs = parser.parse_many(json);
-for (dom::element doc : docs) {
-  cout << doc["foo"] << endl;
+auto json = R"({ "foo": 1 } { "foo": 2 } { "foo": 3 } )"_padded;
```
If it is not already the case, please add a copy of this code example to our tests. Too often we have offered code examples in our documentation that contain minor typos and these are very annoying to newcomers.

lemire replied:
(To be clear, I do not mean that you have offered broken examples... I mean people like me.)

NicolasJiaxin replied:
Added it!

NicolasJiaxin commented:
@lemire For benchmarking, should I aim to benchmark with the same ndjson files than those used for DOM (amazon_cellphones, Kepler, dbSNP)? If so, how were those benchmarked? Also, I only found the amazon_cellphones file and I see that dbSNP benchmark is referenced in #147, but I cannot find the Kepler benchmark.

NicolasJiaxin commented:
And also, I should write include those benchmarks in the benchmark directory, and not just report numbers, right?

lemire commented:
@NicolasJiaxin Give me a few minutes.

lemire commented:
@NicolasJiaxin I looked at  ftp://ftp.ncbi.nih.gov/snp/latest_release/JSON/ It works, but it is somewhat unpleasant to deal with. You can download the files. On the plus side, that's realistic data, stuff that actual people use, but it is also a bit annoying to have to download large files. It will only work locally whereas we would like to have something that we can run easily from the project.

I think that the other thing that @piotte13 did was to convert large CSV files to ndjson. That works also, but it requires you to have the large CSV file.

What we do have (from @piotte13) is jsonexamples/amazon_cellphones.ndjson It is basically one such CSV files, but it can serve us well.

Now  jsonexamples/amazon_cellphones.ndjson is good enough for non-threaded processing (it is 271KB). So we could do a benchmark where we compare our DOM approach and your new code. E.g., here is a fun task given the data. For each row, you have the brand and the rating. So we can go through and compute the average rating per brand. (That's not super meaningful maybe, but it is 'realistic'). We have to be mindful that the first row has to be skipped (since it only contains the labels.)

This could be added to the other benchmarks, like you did a few weeks ago.

It is a bit sad because it won't show up the threaded component (the file being too small).

To do that, we could simply 'replicate' the data. The trick is to load the file, and just keep pasting it in a memory buffer until you have, say, 10MB of data.

So what I suggest is that you start with the first benchmark, with just  jsonexamples/amazon_cellphones.ndjson as-is. We look at the results you are getting. You should not be slower than the DOM equivalent. 

You can load the file like so...
https://github.com/simdjson/simdjson/blob/95b4870e20be5f97d9dcf63b23b1c6f520c366c1/benchmark/find_tweet/find_tweet.h#L14-L16

You are already quite familiar with the document_stream APIs (one could say that you are an expert), and you have created benchmarks using our framework before, so it should be 'easy' for you to do write the code... 

If it works out, then we can extend the benchmark by building a larger file. But I would start with just the same file and see what numbers we are getting. That should allow us to detect a problem, if a problem there.

Moving to the large file should help us tease out the best performance.

Of course, you want to create the large file only once... This can be done like so:

https://github.com/simdjson/simdjson/blob/95b4870e20be5f97d9dcf63b23b1c6f520c366c1/benchmark/large_random/large_random.h#L43-L65

(Notice the 'static' qualifier within the file. It ensures that we create it just once.)

lemire commented:
> And also, I should write include those benchmarks in the benchmark directory, and not just report numbers, right?


Yes. Then we can run the benchmarks on different systems and see what kind of numbers we are getting!!!

lemire commented:
> For benchmarking, should I aim to benchmark with the same ndjson files than those used for DOM (amazon_cellphones, Kepler, dbSNP)? If so, how were those benchmarked? Also, I only found the amazon_cellphones file and I see that dbSNP benchmark is referenced in #147, but I cannot find the Kepler benchmark.

I think I answered this above. The dbSNP are easily accessible but methodologically messy (we can't just include the files as part of our packages and downloading them takes a long time). I do not recall what Kepler was.

Our purpose in this PR is only to document and validate your (excellent!) work, so we do not need to test with lots of different files.

We do need a benchmark because (1) there could be a performance bug in the PR (2) we want to give some indication of the performance to our users.

NicolasJiaxin commented:
OK perfect! I have already try a little bit to setup a amazon_cellphones benchmark by creating a `amazon_cellphones.h`, but I am unable to create a BENCHMARK_TEMPLATE like such: 

https://github.com/simdjson/simdjson/blob/95b4870e20be5f97d9dcf63b23b1c6f520c366c1/benchmark/find_tweet/simdjson_dom.h#L29

Here, the argument `find_tweet` should represent a function?

NicolasJiaxin commented:
I pushed what I did. It is just a preliminary setup, not everything is set up correctly yet to execute the task you described.

lemire commented:
> Here, the argument find_tweet should represent a function? @lemire


It should represent a template as the name `BENCHMARK_TEMPLATE` suggests.

So 

`BENCHMARK_TEMPLATE(f, z)`

will call

`f<z>(benchmark::State &state)`.


This allows you to write the core of the benchmark inside a template (so you avoid ugly copy-paste).

In the case of `find_tweet`, you have...

```C++
template<typename I> simdjson_really_inline static void find_tweet(benchmark::State &state) {
  run_json_benchmark<runner<I>, runner<simdjson_dom>>(state);
}
```

It calls this big thing (run_json_benchmark) there...

https://github.com/simdjson/simdjson/blob/95b4870e20be5f97d9dcf63b23b1c6f520c366c1/benchmark/json_benchmark/run_json_benchmark.h#L19-L105


Don't try to understand this code!!! What you need to understand is that you have to provide a 'runner', it is a thing that has `setup`, `before_run`, ... and so forth. We take two runners, one 'reference' runner (the second one) and a benchmarked runner. This way we get to compare our results with a reference (which avoids benchmarking wrong results). The 'diff' method is what allows you to compare the reference with the benchmarked runner.

We have several example runner throughout the code, including some that load files automagically.

It is intimidating and likely to throw terrible/unreadable errors, but it is not conceptually difficult.

Just ping me if this comment was not enough to make you progress.

NicolasJiaxin commented:
Actually, it was a typo due to laziness of copy-pasting. I forgot to change the name of the template, but your explanation is still useful to understand what happens. Thanks again!

NicolasJiaxin commented:
I will continue working on amazon_cellphones, and report my numbers here as soon as I get them!

lemire commented:
@NicolasJiaxin Note that this is not simdjson-specific. The template in question is from Google Benchmarks which is a widely used framework for benchmarking.

NicolasJiaxin commented:
@lemire I am a little bit stuck benchmarking. The parsing goes fine. The problem comes with the result vector. I have created a `struct brand` which has three fields, `brand_name`, `cumulative_rating`  and `count`. The average rating would just be obtained by dividing `cumulative_rating` by `count` at the end of the parsing. However, each time we iterate to the next document, the `brand_name` of each `brand` in the result vector gets changed to the current document brand name. For example, suppose we have the following document:
```
["asin","brand","title","url","image","rating","reviewUrl","totalReviews","prices"]
[model1, "Nokia", title1, url1, image1, 3, 14, price1]
[model2,"Motorola", title2, url2, image2, 2.9, 7, price2]
````
After the first document, the result vector contains the correct `struct brand` instance that is,
```
brand_name: Nokia
cumulative_rating: 42
count: 14
```
However, when we go to the second document, the brand_name gets changed to "Motor", that is the substring of "Motorola" with same length as "Nokia". This occurs as soon as we change documents, even before we add the "Motorola" data. The values for cumulative_rating and count for each brand are correct at the end document stream though. I have tried copying the string when initializing a `struct brand` instance or making the field `brand_name` const, but it did not work (if you think it works, you can try it. I might have done it wrong). Do you have an idea why it does that?

lemire commented:
@NicolasJiaxin Let me know if the following information helps.

Some of the simdjson library functions will return a `std::string_view` instance which points inside a private string buffer inside the document or the parser instance. A `std::string_view` in C++ is just that... a 'view'... it does not contain the actual string, it just points at it. 

Importantly, creating a `std::string_view` instances bypasses memory allocation. So it is really fast and tight.

If you think about a string in a general in programming, it is a variable-length entity... and it often ends up on the heap (though there are short string optimizations). It can be very expensive... See this blog post... [Performance tip: constructing many non-trivial objects is slow](https://lemire.me/blog/2020/08/08/performance-tip-constructing-many-non-trivial-objects-is-slow/).

So whenever possible, you should use something like a `std::string_view`.

However, a `std::string_view` instance has a life cycle that is tied to the owner of the data. Someone, somewhere, must have allocated the memory and must be holding it for you.

In the case of simdjson, this entity is almost surely reusing this memory region from document to document.

So if you grab a `string_view` instance from a document in simdjson, unless otherwise stated, this string view only lives for the duration of the current document instance or the equivalent.

What if you need to keep it for longer? Then you must copy it. You can do so rather easily in C++ by copying it to a `std::string`...

```
std::string my_copy(my_string_view.begin(),my_string_view.end()); 
```

As my blog post above suggests, you really ought to avoid that if you can. If the `brand` is a key, then you could treat it as such... (comment to be followed by actual code)

lemire commented:
@NicolasJiaxin Visual Studio is doing its thing...

```
simdjson\benchmark\amazon_cellphones/simdjson_dom.h(34,1): warning C4244: '+=': conversion from 'T' to 'size_t', possible loss of data [D:\a\simdjson\simdjson\build\benchmark\bench_ondemand.vcxproj]
```

Try to use explicit casts.

lemire commented on benchmark/amazon_cellphones/amazon_cellphones.h:
```diff
@@ -0,0 +1,75 @@
+
+template<typename StringType>
+struct brand {
+    const StringType brand_name;
```
If `StringType` is a `std::string_view`, you have to ensure that the `std::string_view` remains valid for the duration of the life of your object.

See 'Notes' in https://en.cppreference.com/w/cpp/string/basic_string_view

> It is the programmer's responsibility to ensure that std::string_view does not outlive the pointed-to character array:


Note that a `string_view` does not actually own the string content. It only points at it.

lemire commented on benchmark/amazon_cellphones/simdjson_dom.h:
```diff
@@ -0,0 +1,49 @@
+      auto doc = *i;
+      auto x = result.begin();
+      while(true) {
+        if (x == result.end()) {
```
Here `result` is a vector which accumulates all of the data parsed. If you only want an aggregated result, then it is wasteful.

(I will provide a code example soon.)

lemire commented:
@NicolasJiaxin Ok. 

The following will have sane performance if your strings are relatively short so that the "short string" optimization of the standard library kicks in...

```C++
struct basic_accumulator {
    std::map<std::string,size_t> content; 

    void consume(std::string_view view) {
        std::string copy(view);
        auto i = content.find(copy);
        if(i == content.end()) { // key is not found
          content.emplace(copy, 1);
        } else {
          i->second++;
        }
    }


};

```

The following might be more performant in some instances...

```C++
struct view_accumulator {
    std::map<std::string_view,size_t> content; // keys point at string_buffer
    std::vector<char> string_buffer; // owns the string content

    void consume(std::string_view view) {
        auto i = content.find(view);
        if(i == content.end()) { // key is not found
          size_t where_at = string_buffer.size();
          string_buffer.insert(string_buffer.end(), view.begin(), view.end());
          std::string_view permanent_view(&string_buffer[where_at], view.size());
          content.emplace(permanent_view, 1);
        } else {
          i->second++;
        }
    }

    size_t result() {
        return content.size();
    }
};

```

Of course, my 'accumulators' do not do exactly what you have in mind but they serve to illustrate the concepts.

[I posted a full program online](https://github.com/lemire/Code-used-on-Daniel-Lemire-s-blog/blob/master/extra/viewstring/views.cpp).

lemire commented:
@NicolasJiaxin In your case, because you are dealing with short strings, I expect that `basic_accumulator` is the model to follow. So you copy the `string_view` instances to local `std::string` instances but because they are small, there is no dynamic heap allocation... hence, it is pretty fast.

NicolasJiaxin commented:
@lemire it worked, thank you!! I used the model of the `basic_accumulator`. So, if I understand well, this is fast because the strings are small, and hence don't need to be allocated on the heap. What happens if the string slightly larger than the "short string" optimization? Since you talked about the standard library, I guess that means that there exists other StringType implementations with better optimizations for larger sizes.

lemire commented:
@NicolasJiaxin That's correct.

What constitutes a short string, and how well it is implemented is almost surely up to the authors of the standard library... so you should expect large differences between Linux (gnu), macOS/freeBSD and Visual Studio. It is likely that some of these libraries offer some tuning knobs you can play with.

There is a trade-off involved. Supporting very large 'short strings' would potentially be wasteful... since it is almost surely done using a static (fixed) container. 

If your application ends up using larger strings, then the performance will suffer, possibly quite a bit (e.g., it could become 10x slower). In our case, we know our data, so we know it is not an issue. I would bet that the strings are considered 'small' by most frameworks.

Importantly, C++ makes it easy for you to write your own string type. In fact, you have been using a padded_string type, which is our own type. This is not at all theoretical. Many frameworks rely on their own string types... which is one argument for simdjson not to lock the users down with a specific storage mechanism.

So you could just 'nail' the problem and write your own string class for this problem. In this case, it would be absurd, but if the problem because important enough, it would be an option. 

In simdjson, we simply avoid the problem and go with string_view, and leave it up the users to decide how to store the strings... Because we think it is best to give people options.

The price to pay is that people have to understand how to use string_view instances...

NicolasJiaxin commented:
Here are some numbers:

<table>
<center><b> Single amazon_cellphones document (271KB) </b></center>
<tr><td>

|Benchmark|throughput|best_branch_miss|
|-----|-----|-----|
|amazon_cellphones<simdjson_dom>/manual_time|1.66GB/s|2.312k|
|amazon_cellphones<simdjson_ondemand>/manual_time|2.02GB/s|1.555k|
</td></tr>
</table>

From DOM to On Demand, this represents an increase in throughput of 22% which is a good sign that document_stream performs well, but I cannot say myself if this is above/below the expectation. Also, keep in mind that this benchmark is not large enough to properly benchmark the threaded versions of DOM and On Demand. More benchmarks will be done for that, this is a preliminary result.

lemire commented:
Congratulations!

These numbers look good to me.

We do want a larger input to test the multithreaded component… but it is clear already that you got the core component to work properly! Not an easy task!

NicolasJiaxin commented:
I have now added a large benchmark(38 copies of amazon_cellphones that spans ~10MB) to test the threaded version. Here are the numbers that I get:

<table>
<center><b> Large amazon_cellphones document (10MB) </b></center>
<tr><td>

|Benchmark|throughput|
|-----|-----|
|large_amazon_cellphones<simdjson_dom>/manual_time|2.09GB/s|
|large_amazon_cellphones<simdjson_ondemand>/manual_time|2.31GB/s|
</td></tr>
</table> 

As expected, both DOM and On Demand showed increased throughput, and as documented, DOM greatly benefits from parallelization (+26%). However, On Demand did not show an increase as substantial as DOM, but still an increase of 14% in throughput. We might want to revisit the threaded implementation of On Demand if this is not deemed satisfactory.

lemire commented:
@NicolasJiaxin This is a good validation.

lemire commented:
I think we will merge, if it passes our tests. The code of the benchmarks is nice.

lemire commented:
@NicolasJiaxin We seem to have a few CI failures. When you get around to it, could you assess them?

NicolasJiaxin commented:
Yes, no problem.

lemire commented:
@NicolasJiaxin The failure seems to be specific to one test `ondemand_readme_examples`.

NicolasJiaxin commented:
It should be fixed now @lemire. I fixed the test and updated the documentation accordingly. Like in DOM, I also added a `parser.threaded` field that is enabled by default when SIMDJSON_THREADS_ENABLED is on.

lemire commented:
So far, CI agrees with you!

NicolasJiaxin commented:
It's ready @lemire !

lemire commented:
Yep. I'll have a quick look this afternoon, but I expect to merge!

lemire commented:
Merging. This is quite a big piece!
